cirq/testing/circuit_compare.py

```diff
@@ -245,6 +245,113 @@ def assert_has_diagram(
     )
 
 
+def assert_has_consistent_apply_unitary(
+        val: Any,
+        *,
+        qubit_count: Optional[int] = None) -> None:
+    """Tests whether a value's _apply_unitary_to_tensor_ is correct.
+
+    Contrasts the effects of the value's `_apply_unitary_to_tensor_` with the
+    matrix returned by the value's `_unitary_` method.
+
+    Args:
+        val: The value under test. Should have a `__pow__` method.
+        qubit_count: Usually inferred. The number of qubits the value acts on.
+            This argument isn't needed if the gate has a unitary matrix or
+            implements `cirq.SingleQubitGate`/`cirq.TwoQubitGate`/
+            `cirq.ThreeQubitGate`.
+    """
+
+    expected = protocols.unitary(val, default=None)
+    if qubit_count is not None:
+        n = qubit_count
+    elif expected is not None:
+        n = expected.shape[0].bit_length() - 1
+    else:
+        n = _infer_qubit_count(val)
+
+    eye = np.eye(2 << n, dtype=np.complex128).reshape((2,) * (2 * n + 2))
+    actual = protocols.apply_unitary_to_tensor(
+        val=val,
+        target_tensor=eye,
+        available_buffer=np.ones_like(eye) * float('nan'),
+        axes=list(range(1, n + 1)),
+        default=None)
+
+    # If you don't have a unitary, you shouldn't be able to apply a unitary.
```

Comment:
I'm not sure I agree with this; I think it's okay to define `apply_unitary_effect_to_state` without defining `unitary`.

Reply:
I went back and forth on this as well, but ultimately I came down on "it's easier to make it looser later than stricter later". I suspect this may be resolved by a hypothetical possibly-maybe future change that implements `_unitary_` in terms of `_apply_unitary_to_tensor_` when only the latter is present.